crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -1180,6 +1181,252 @@ impl<'w> EntityWorldMut<'w> {
+    ///
+    /// #[derive(Component)]
+    /// #[require(Y)]
+    /// struct X;
+    ///
+    /// #[derive(Component, Default)]
+    /// #[require(Z)]
+    /// struct Y;
+    ///
+    /// #[derive(Component, Default)]
+    /// struct Z;
+    ///
+    /// #[derive(Component)]
+    /// #[require(Z)]
+    /// struct W;
+    ///
+    /// let mut world = World::new();
+    ///
+    /// // Spawn an entity with X, Y, Z, and W components
+    /// let entity = world.spawn((X, W)).id();
+    ///
+    /// // Initial component tree:
+    /// //  X
+    /// //   \
+    /// //    Y   W
+    /// //     \ /
+    /// //      Z
+    ///
+    /// assert!(world.entity(entity).contains::<X>());
+    /// assert!(world.entity(entity).contains::<Y>());
+    /// assert!(world.entity(entity).contains::<Z>());
+    /// assert!(world.entity(entity).contains::<W>());
+    ///
+    /// // Remove X and its unused requirements
+    /// world.entity_mut(entity).require_recursive_remove::<X>();
+    ///
+    /// // Resulting component tree:
+    /// //     W
+    /// //    /
+    /// //   Z
+    ///
+    /// assert!(!world.entity(entity).contains::<X>());
+    /// assert!(!world.entity(entity).contains::<Y>());
+    /// assert!(world.entity(entity).contains::<Z>());
+    /// assert!(world.entity(entity).contains::<W>());
+    /// ```
+    /// Note: It will note delete components in [`Bundle`] if some components require bundle components. Use
+    ///
+    pub fn require_recursive_remove<T: Bundle>(&mut self) -> &mut Self {
+        let storages = &mut self.world.storages;
+        let components = &mut self.world.components;
+        let bundle = self.world.bundles.init_info::<T>(components, storages);
+
+        // SAFETY: the `BundleInfo` is initialized above
+        // We make clone of components array to not lock immutable access to world
+        let contributed_components = unsafe {
+            self.world
+                .bundles
+                .get_unchecked(bundle)
+                .contributed_components()
+                .to_vec()
+        };
+
+        let old_location = self.location;
+        let old_archetype = &mut self.world.archetypes[old_location.archetype_id];
+
+        let remaining_components = old_archetype
+            .components()
+            .filter(|c| !contributed_components.contains(c))
+            .collect::<Vec<_>>();
+
+        let remaining_bundle = self
+            .world
+            .bundles
+            .init_dynamic_info(components, &remaining_components);
+
+        // SAFETY: the `BundleInfo` is initialized above
+        let remaining_bundle_info = unsafe { self.world.bundles.get_unchecked(remaining_bundle) };
+
+        // Remove components required by remaining components from contributed_components of bundle T
+        let to_delete = contributed_components
+            .iter()
+            .filter(|c| !remaining_bundle_info.contributed_components().contains(c))
+            .copied()
+            .collect::<Vec<_>>();
```

Comment:
I feel like the logic is becoming complex enough that it might need a bit more comments. From what I see the process is essentially this:

1. gather all the components that the `T` bundle would contribute (`T`'s components, their required components, required components of those required components, and so on...) in `contributed_components`;
    - these might however contains components that are required by other existing components, so they can't be removed immediately.
2. gather all the remaining components in `remaining_components`, that is all the components in this archetypes that are not included in 1.;
3. find all the components that these components would contribute with `remaining_bundle_info.contributed_components()`, these are those that we want to keep;
4. thus we have to remove all the others, which are computed in `to_delete`.

Might be worth comment where each of these steps happens.

Something similar, although even more complex, seems to be happening in `require_descendant_remove`, so it would be nice to add some comments there as well.